azext_iot/iothub/providers/device_messaging.py

```diff
@@ -0,0 +1,527 @@
+                cancellation_token.set()
+
+    def device_upload_file(
+        self,
+        file_path: str,
+        content_type: str,
+    ):
+        from azext_iot.sdk.iothub.device.models import FileUploadCompletionStatus
+
+        if not exists(file_path):
+            raise FileOperationError('File path "{}" does not exist!'.format(file_path))
+
+        content = read_file_content(file_path)
+        file_name = basename(file_path)
+
+        try:
+            upload_meta = self.device_sdk.device.create_file_upload_sas_uri(
+                device_id=self.device_id, blob_name=file_name, raw=True
+            ).response.json()
+            storage_endpoint = "{}/{}/{}{}".format(
+                upload_meta["hostName"],
+                upload_meta["containerName"],
+                upload_meta["blobName"],
+                upload_meta["sasToken"],
+            )
+            completion_status = FileUploadCompletionStatus(
+                correlation_id=upload_meta["correlationId"], is_success=True
+            )
+            upload_response = self.device_sdk.device.upload_file_to_container(
+                storage_endpoint=storage_endpoint,
+                content=content,
+                content_type=content_type,
+            )
+            completion_status.status_code = upload_response.status_code
+            completion_status.status_reason = upload_response.reason
+
+            return self.device_sdk.device.update_file_upload_status(
+                device_id=self.device_id, file_upload_completion_status=completion_status
+            )
+        except CloudError as e:
+            handle_service_exception(e)
+
+    def _d2c_get_device_auth_props(
+        self,
+        symmetric_key: Optional[str] = None,
+        certificate_file: Optional[str] = None,
+        key_file: Optional[str] = None,
+        passphrase: Optional[str] = None
+    ):
+        if symmetric_key:
+            return {
+                "hub": self.target["entity"],
+                "deviceId": self.device_id,
+                "authentication": {
+                    "type": DeviceAuthApiType.sas.value,
+                    "symmetricKey": {
+                        "primaryKey": symmetric_key
+                    }
+                }
+            }
+        elif (certificate_file and key_file):
+            # custom device structure to hold needed info
+            # Note that here signed vs ca doesnt matter. CA will need a verified cert
+            # (stuff from dps enrollment group certs)
+            # Note that for the CA device the subject of the cert must be the device_id
+            return {
+                "deviceId": self.device_id,
+                "authentication": {
+                    "type": DeviceAuthApiType.selfSigned.value,
+                    "x509_files": {
+                        "certificateFile": certificate_file,
+                        "keyFile": key_file,
+                        "passphrase": passphrase
```

Comment:
Can passphrase be an empty string or None?

Reply:
either works